src/objective/cuda/cuda_binary_objective.cu

```diff
@@ -0,0 +1,227 @@
+                                                     const data_size_t num_data, const double sigmoid) {
+  const double suml = *out_cuda_sum_labels;
+  const double sumw = USE_WEIGHT ? *out_cuda_sum_weights : static_cast<double>(num_data);
+  double pavg = suml / sumw;
+  pavg = min(pavg, 1.0 - kEpsilon);
+  pavg = max(pavg, kEpsilon);
+  const double init_score = log(pavg / (1.0f - pavg)) / sigmoid;
+  *out_cuda_sum_weights = pavg;
+  *out_cuda_sum_labels = init_score;
+}
+
+void CUDABinaryLogloss::LaunchBoostFromScoreKernel() const {
+  const int num_blocks = (num_data_ + CALC_INIT_SCORE_BLOCK_SIZE_BINARY - 1) / CALC_INIT_SCORE_BLOCK_SIZE_BINARY;
+  if (ova_class_id_ == -1) {
+    if (cuda_weights_ == nullptr) {
+      BoostFromScoreKernel_1_BinaryLogloss<false, false><<<num_blocks, CALC_INIT_SCORE_BLOCK_SIZE_BINARY>>>
+        (cuda_label_, num_data_, cuda_boost_from_score_, cuda_sum_weights_, cuda_weights_, ova_class_id_);
+    } else {
+      BoostFromScoreKernel_1_BinaryLogloss<false, true><<<num_blocks, CALC_INIT_SCORE_BLOCK_SIZE_BINARY>>>
+        (cuda_label_, num_data_, cuda_boost_from_score_, cuda_sum_weights_, cuda_weights_, ova_class_id_);
+    }
+  } else {
+    if (cuda_weights_ == nullptr) {
+      BoostFromScoreKernel_1_BinaryLogloss<true, false><<<num_blocks, CALC_INIT_SCORE_BLOCK_SIZE_BINARY>>>
+        (cuda_label_, num_data_, cuda_boost_from_score_, cuda_sum_weights_, cuda_weights_, ova_class_id_);
+    } else {
+      BoostFromScoreKernel_1_BinaryLogloss<true, true><<<num_blocks, CALC_INIT_SCORE_BLOCK_SIZE_BINARY>>>
+        (cuda_label_, num_data_, cuda_boost_from_score_, cuda_sum_weights_, cuda_weights_, ova_class_id_);
+    }
+  }
+  SynchronizeCUDADevice(__FILE__, __LINE__);
+  if (cuda_weights_ == nullptr) {
+    BoostFromScoreKernel_2_BinaryLogloss<false><<<1, 1>>>(cuda_boost_from_score_, cuda_sum_weights_, num_data_, sigmoid_);
+  } else {
+    BoostFromScoreKernel_2_BinaryLogloss<true><<<1, 1>>>(cuda_boost_from_score_, cuda_sum_weights_, num_data_, sigmoid_);
+  }
+  SynchronizeCUDADevice(__FILE__, __LINE__);
+}
+
+template <bool USE_LABEL_WEIGHT, bool USE_WEIGHT, bool IS_OVA>
+__global__ void GetGradientsKernel_BinaryLogloss(const double* cuda_scores, const label_t* cuda_labels,
+  const double* cuda_label_weights, const label_t* cuda_weights, const int ova_class_id,
+  const double sigmoid, const data_size_t num_data,
+  score_t* cuda_out_gradients, score_t* cuda_out_hessians) {
+  const data_size_t data_index = static_cast<data_size_t>(blockDim.x * blockIdx.x + threadIdx.x);
+  if (data_index < num_data) {
+    const label_t cuda_label = static_cast<int>(cuda_labels[data_index]);
+    const int label = IS_OVA ? (cuda_label == ova_class_id ? 1 : -1) : (cuda_label > 0 ? 1 : -1);
+    const double response = -label * sigmoid / (1.0f + exp(label * sigmoid * cuda_scores[data_index]));
+    const double abs_response = fabs(response);
+    if (!USE_WEIGHT) {
+      if (USE_LABEL_WEIGHT) {
```

Comment:
should we use `constexpr` for these template conditions ?

Reply:
constexpr if is only available with c++17 or higher version.